coatrack/coatrack
# Facilitate dockerized CoatRack deployment, remove dependencies on external Git service and external CoatRack gateway jar

### Bugfixes and features
This PR covers several issues and features:
* The Gateway creation was fixed by getting rid the external bitbucket service dependency, see #186. The Gateway configuration files are now stored locally when IDE/CLI is uesed or within a docker volume when docker-compose is used for deployment. 
* The proxy jar is now stored locally as well instead of pulling it from an external server, see #167. 
* For developers: CoatRack Web Application and Config Server can be started via IDE or CLI without requiring any pre-configuration or active profiles. They work out of the box now. The Web Application will use the proxy jar file generated by the maven build process. See #184 and #172.
* A full CoatRack instance with DB can theoretically easily be deployed using the CoatRack images from Dockerhub as soon as the images of this branch are pushed, see README and #162. Because of that, this feature could not be tested yet. We should discuss that first.

I also added some documentation as new sections in README.md: "Installation using Docker Images" (not working because this PR is not merged yet and master has gateway creation problems because it still depends on the broken bitbucket service) and "Operations from Scratch".

### How to test this PR
* First approach: If not existent, build CoatRack to create the file `/spring-boot/proxy/target/<proxy>.jar`. This can be achieved by executing the `build.sh` file in the project directory. Please try to run CoatRack via simply starting the Web Application/Admin and Config server via IDE or CLI without any active profile and check if it works for you.
* Second approach: Go to the "docker" folder and run `bash build-and-deploy-images-locally.sh'. A full, dockerized CoatRack instance should be set up.

Especially test if the gateway creation, download, deployment and service-call-via-api-key feature are working correctly for you.

closes #167, #184, #172, #162, #186 

### TODO for me before merging
 - [x] Add a script that builds starts admin and one that starts config-server. Starting a component should not depend on the IDE.
 - [x]  Discuss if I can push the images of this branch to the Dockerhub repository to test the use cases from the section of the  README.md: "Installation using Docker Images".
 - [x] The Pipeline failed. Probably because I moved some scripts to the `docker` folder. To be fixed. -> I solved that by putting this folder from the "target" directory to the project directory.
